File: TaskCLI/cpp/src/storage/json_store.cpp
// src/storage/json_store.cpp

#include "storage/json_store.h"
#include <fstream>
#include <algorithm>

using json = nlohmann::json;

Storage::Storage(const std::string& data_dir) 
    : data_dir_(data_dir), next_task_id_(1), next_user_id_(1) {
    
    std::filesystem::create_directories(data_dir_);
    load();
}

void Storage::load() {
    std::ifstream file(data_dir_ + "/data.json");
    if (file.is_open()) {
        try {
            json j;
            file >> j;
            
            for (const auto& u : j["users"]) {
                users_.push_back(std::make_shared<User>(User::from_json(u)));
            }
            
            for (const auto& t : j["tasks"]) {
                // 简化的反序列化
                auto task = std::make_shared<Task>(t["id"], t["title"], t["user_id"]);
                task->priority = Task::priority_from_string(t["priority"]);
                task->status = Task::status_from_string(t["status"]);
                tasks_.push_back(task);
            }
            
            next_user_id_ = j.value("next_user_id", 1);
            next_task_id_ = j.value("next_task_id", 1);
        } catch (...) {
            // 如果加载失败，使用默认数据
        }
    }
    
    if (users_.empty()) {
        users_.push_back(std::make_shared<User>(User(1, "default")));
        next_user_id_ = 2;
    }
}

void Storage::save() {
    json j;
    
    for (const auto& u : users_) {
        j["users"].push_back(u->to_json());
    }
    
    for (const auto& t : tasks_) {
        json task_json;
        task_json["id"] = t->id;
        task_json["title"] = t->title;
        task_json["priority"] = t->priority_to_string();
        task_json["status"] = t->status_to_string();
        task_json["user_id"] = t->user_id;
        task_json["tags"] = t->tags;
        j["tasks"].push_back(task_json);
    }
    
    j["next_user_id"] = next_user_id_;
    j["next_task_id"] = next_task_id_;
    
    std::ofstream file(data_dir_ + "/data.json");
    file << j.dump(2);
}

std::vector<std::shared_ptr<User>> Storage::get_users() {
    return users_;
}

std::shared_ptr<User> Storage::add_user(const std::string& name) {
    auto user = std::make_shared<User>(User(next_user_id_++, name));
    users_.push_back(user);
    save();
    return user;
}

void Storage::delete_user(uint32_t id) {
    users_.erase(std::remove_if(users_.begin(), users_.end(),
        [id](const auto& u) { return u->id == id; }), users_.end());
    
    tasks_.erase(std::remove_if(tasks_.begin(), tasks_.end(),
        [id](const auto& t) { return t->user_id == id; }), tasks_.end());
    
    save();
}

std::optional<uint32_t> Storage::get_user_id(const std::string& name) {
    for (const auto& u : users_) {
        if (u->name == name) return u->id;
    }
    return std::nullopt;
}

std::vector<std::shared_ptr<Task>> Storage::get_tasks() {
    return tasks_;
}

std::shared_ptr<Task> Storage::get_task(uint32_t id) {
    for (const auto& t : tasks_) {
        if (t->id == id) return t;
    }
    return nullptr;
}

std::shared_ptr<Task> Storage::add_task(std::shared_ptr<Task> task) {
    task->id = next_task_id_++;
    tasks_.push_back(task);
    save();
    return task;
}

void Storage::update_task(std::shared_ptr<Task> task) {
    for (auto& t : tasks_) {
        if (t->id == task->id) {
            t = task;
            break;
        }
    }
    save();
}

void Storage::delete_task(uint32_t id) {
    tasks_.erase(std::remove_if(tasks_.begin(), tasks_.end(),
        [id](const auto& t) { return t->id == id; }), tasks_.end());
    save();
}

std::vector<std::shared_ptr<Task>> Storage::get_tasks_by_user(uint32_t user_id) {
    std::vector<std::shared_ptr<Task>> result;
    for (const auto& t : tasks_) {
        if (t->user_id == user_id) result.push_back(t);
    }
    return result;
}

std::vector<std::shared_ptr<Task>> Storage::get_tasks_by_status(Status status) {
    std::vector<std::shared_ptr<Task>> result;
    for (const auto& t : tasks_) {
        if (t->status == status) result.push_back(t);
    }
    return result;
}

std::vector<std::shared_ptr<Task>> Storage::get_tasks_by_tag(const std::string& tag) {
    std::vector<std::shared_ptr<Task>> result;
    for (const auto& t : tasks_) {
        if (t->has_tag(tag)) result.push_back(t);
    }
    return result;
}

std::vector<std::shared_ptr<Task>> Storage::search_tasks(const std::string& keyword) {
    std::vector<std::shared_ptr<Task>> result;
    for (const auto& t : tasks_) {
        if (t->matches(keyword)) result.push_back(t);
    }
    return result;
}

Storage::Stats Storage::get_stats(std::optional<uint32_t> user_id) {
    Stats s{0, 0, 0, 0, 0};
    
    for (const auto& t : tasks_) {
        if (user_id.has_value() && t->user_id != user_id.value()) continue;
        
        s.total++;
        switch (t->status) {
            case Status::Todo: s.todo++; break;
            case Status::InProgress: s.in_progress++; break;
            case Status::Done: s.done++; break;
            case Status::Cancelled: s.cancelled++; break;
        }
    }
    
    return s;
}

File: TaskCLI/rust/src/storage/json_store.rs
// src/storage/json_store.rs - JSON 文件存储
// 展示 Rust 的文件 I/O、错误处理、Result

use anyhow::{Context, Result};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::fs;
use std::path::PathBuf;

use crate::model::{Task, User};

/// 应用数据容器
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct AppData {
    pub users: Vec<User>,
    pub tasks: Vec<Task>,
    pub next_user_id: u32,
    pub next_task_id: u32,
}

impl AppData {
    /// 创建新的应用数据
    pub fn new() -> Self {
        Self {
            users: Vec::new(),
            tasks: Vec::new(),
            next_user_id: 1,
            next_task_id: 1,
        }
    }

    /// 添加默认用户
    pub fn init_default_user(&mut self) {
        if self.users.is_empty() {
            self.users.push(User::new(1, "default".to_string()));
        }
    }
}

/// 存储管理器
pub struct Storage {
    data_dir: PathBuf,
    data: AppData,
}

impl Storage {
    /// 创建新的存储管理器
    pub fn new(data_dir: PathBuf) -> Result<Self> {
        // 确保目录存在
        fs::create_dir_all(&data_dir).context("Failed to create data directory")?;

        // 加载数据
        let data = Self::load_from_file(&data_dir)?;

        Ok(Self { data_dir, data })
    }

    /// 从文件加载数据
    fn load_from_file(data_dir: &PathBuf) -> Result<AppData> {
        let data_file = data_dir.join("data.json");

        if data_file.exists() {
            let content = fs::read_to_string(&data_file).context("Failed to read data file")?;

            let mut data: AppData =
                serde_json::from_str(&content).context("Failed to parse data file")?;

            // 初始化默认用户
            data.init_default_user();

            Ok(data)
        } else {
            // 创建默认数据
            let mut data = AppData::new();
            data.init_default_user();

            // 保存到文件
            let content = serde_json::to_string_pretty(&data)?;
            fs::write(&data_file, content)?;

            Ok(data)
        }
    }

    /// 保存数据到文件
    pub fn save(&self) -> Result<()> {
        let data_file = self.data_dir.join("data.json");
        let content = serde_json::to_string_pretty(&self.data)?;
        fs::write(&data_file, content)?;
        Ok(())
    }

    // ==================== 用户操作 ====================

    /// 获取所有用户
    pub fn get_users(&self) -> &[User] {
        &self.data.users
    }

    /// 添加用户
    pub fn add_user(&mut self, name: String) -> Result<User> {
        let user = User::new(self.data.next_user_id, name);
        self.data.next_user_id += 1;
        self.data.users.push(user.clone());
        self.save()?;
        Ok(user)
    }

    /// 删除用户
    pub fn delete_user(&mut self, id: u32) -> Result<()> {
        self.data.users.retain(|u| u.id != id);
        // 同时删除用户的任务
        self.data.tasks.retain(|t| t.user_id != id);
        self.save()?;
        Ok(())
    }

    /// 获取用户 ID（按名称）
    pub fn get_user_id(&self, name: &str) -> Option<u32> {
        self.data
            .users
            .iter()
            .find(|u| u.name == name)
            .map(|u| u.id)
    }

    // ==================== 任务操作 ====================

    /// 获取所有任务
    pub fn get_tasks(&self) -> &[Task] {
        &self.data.tasks
    }

    /// 获取任务（按 ID）
    pub fn get_task(&self, id: u32) -> Option<&Task> {
        self.data.tasks.iter().find(|t| t.id == id)
    }

    /// 添加任务
    pub fn add_task(&mut self, task: Task) -> Result<Task> {
        let id = self.data.next_task_id;
        let mut task = task;
        task.id = id;
        self.data.next_task_id += 1;
        self.data.tasks.push(task.clone());
        self.save()?;
        Ok(task)
    }

    /// 更新任务
    pub fn update_task(&mut self, task: Task) -> Result<()> {
        if let Some(existing) = self.data.tasks.iter_mut().find(|t| t.id == task.id) {
            *existing = task;
            self.save()?;
            Ok(())
        } else {
            Err(anyhow::anyhow!("Task not found"))
        }
    }

    /// 删除任务
    pub fn delete_task(&mut self, id: u32) -> Result<()> {
        self.data.tasks.retain(|t| t.id != id);
        self.save()?;
        Ok(())
    }

    /// 按用户过滤任务
    pub fn get_tasks_by_user(&self, user_id: u32) -> Vec<&Task> {
        self.data
            .tasks
            .iter()
            .filter(|t| t.user_id == user_id)
            .collect()
    }

    /// 按状态过滤任务
    pub fn get_tasks_by_status(&self, status: crate::model::Status) -> Vec<&Task> {
        self.data
            .tasks
            .iter()
            .filter(|t| t.status == status)
            .collect()
    }

    /// 按标签过滤任务
    pub fn get_tasks_by_tag(&self, tag: &str) -> Vec<&Task> {
        self.data.tasks.iter().filter(|t| t.has_tag(tag)).collect()
    }

    /// 搜索任务
    pub fn search_tasks(&self, keyword: &str) -> Vec<&Task> {
        self.data
            .tasks
            .iter()
            .filter(|t| t.matches(keyword))
            .collect()
    }

    /// 获取任务统计
    pub fn get_stats(&self, user_id: Option<u32>) -> TaskStats {
        let tasks: Vec<&Task> = match user_id {
            Some(id) => self.data.tasks.iter().filter(|t| t.user_id == id).collect(),
            None => self.data.tasks.iter().collect(),
        };

        TaskStats {
            total: tasks.len(),
            todo: tasks
                .iter()
                .filter(|t| t.status == crate::model::Status::Todo)
                .count(),
            in_progress: tasks
                .iter()
                .filter(|t| t.status == crate::model::Status::InProgress)
                .count(),
            done: tasks
                .iter()
                .filter(|t| t.status == crate::model::Status::Done)
                .count(),
            cancelled: tasks
                .iter()
                .filter(|t| t.status == crate::model::Status::Cancelled)
                .count(),
            by_priority: self::count_by_priority(&tasks),
        }
    }
}

fn count_by_priority(tasks: &[&Task]) -> HashMap<String, usize> {
    let mut counts = HashMap::new();
    for task in tasks {
        *counts.entry(task.priority.to_string()).or_insert(0) += 1;
    }
    counts
}

/// 任务统计
#[derive(Debug)]
pub struct TaskStats {
    pub total: usize,
    pub todo: usize,
    pub in_progress: usize,
    pub done: usize,
    pub cancelled: usize,
    pub by_priority: HashMap<String, usize>,
}
